### DISCUSSION TITLE
Using parallel meep

### DISCUSSION THREAD

**Discussion by LaraCouronne** [NanoComp/meep#2437] [Q&A] [answered]
Hello again !
A coworker and I are having an issue with parallel meep.

Our simulations consist of one source propagating through a (quite large) given structure. As the execution of that kind of simulations easily reach hours, we want it to be processed by several cores to make it faster.

We run our codes through the terminal with commands such as : 
```
mpirun -np 4 python my_meep_code.py 
```

However, if we call n core in our command, we get n time the outputs : for the number of plots shown as well as for the number of data saved in files (now I do use `if mp.am_master():` to ensure only one plot is made or one set of data is saved, however it does not ensure that parallel meep is actually parallelizing the process the way we want).

Furthermore, it seems like the time isn't linearly reducing with the number of cores requested, which leads us to think that this command might be running the same code on each core instead of divising it.

How should we use parallel meep to make it efficient ? If the simulation is actually being split on the cores properly, how come the outputs are obtained several times and the time not reduced linearly ?

Best regards !
Lara Couronné

**Comment by Andeloth** [comment, marked as answer]
Meep runs the python script you write on every process you start, only the simulation itself is split between processes. Therefore if you have any print statements or save any data without mp.am_master(), then it will run it on all processes.

As for the speedup, did you install the parallel meep package with `conda install -c conda-forge pymeep="*=mpi_mpich*"` or did you just run `conda install -c conda-forge pymeep`? The first one is needed for parallel meep. If you installed the correct meep version then this will print when meep runs:

`Using MPI version 3.1, 2 processes`.

If you did install parallel meep, then have you checked the simulation step time? Run `mp.verbosity(3)` in your script and then it should print out the time a simulation step takes like this:

```
on time step 81 (time=1.35), 0.0505804 s/step
DFT fields decay(t = 1.57): 0.0000e+00
on time step 158 (time=2.63333), 0.0520784 s/step
DFT fields decay(t = 3.12): 1.0000e+00
```

If you're running the simulation on multiple cores then the time/step will go down with each additional process.

**Reply by LaraCouronne** [reply]
I did install the parallel meep package, yes. I checked the time/step and it seems to be decreasing when I use multiple cores, so I guess that it is working correctly and it is juste the simulation that is really big. Thank you !
